deltacat/compute/compactor/model/hash_bucket_result.py

```diff
@@ -0,0 +1,9 @@
+from dataclasses import dataclass
+
+import numpy as np
+
+
+@dataclass
+class HashBucketResult:
+    hash_bucket_group_to_obj_id: np.ndarray
+    hb_record_count: int
```

Comment:
If we're talking about just updating `hb_record_count` to a NumPy primitive int, the improvement here could be negligible. We could also just include it as a part of #128.

Reply:
Maybe its worth benchmarking. 8000 * 8000 int is nearly 512 MB (considering 8000 hb each with 8000 envelopes), using numpy will allow zero copy transfer without (de)serialization. I am also skeptical about returning `@dataclass` here as it may deteriorate performance. From the Ray documentation, they optimize storing python collections by converting to arrow memory format but I didn't see similar for arbitrary classes. At the end of the day, it may depend on how cloudpickle works. 

Quoting from [this](https://docs.ray.io/en/latest/ray-core/objects/serialization.html#serialization-notes):

> Whenever possible, use numpy arrays or Python collections of numpy arrays for maximum performance.

Reply:
Yeah it could be worth benchmarking. FWIW though, Ray Datasets itself uses `@dataclass` annotations for `BlockMetadata`: https://github.com/ray-project/ray/blob/releases/2.3.1/python/ray/data/block.py#L211